pycalphad/codegen/callables.py

```diff
@@ -0,0 +1,173 @@
+import pycalphad.variables as v
+from pycalphad import Model
+from pycalphad.codegen.sympydiff_utils import build_functions
+from pycalphad.core.utils import get_pure_elements, unpack_components, unpack_kwarg
+from pycalphad.core.phase_rec import PhaseRecord_from_cython
+from sympy import Symbol
+import numpy as np
+import operator
+from itertools import repeat
+from collections import defaultdict
+
+
+def wrap_symbol(obj):
+    if isinstance(obj, Symbol):
+        return obj
+    else:
+        return Symbol(obj)
+
+def build_callables(dbf, comps, phases, model=None, parameters=None, callables=None,
+                    output='GM', build_gradients=True, verbose=False):
+    """
+    Create dictionaries of callable dictionaries and PhaseRecords.
+
+    Parameters
+    ----------
+    dbf : Database
+        A Database object
+    comps : list
+        List of component names
+    phases : list
+        List of phase names
+    model : dict or type
+        Dictionary of {phase_name: Model subclass} or a type corresponding to a
+        Model subclass. Defaults to ``Model``.
+    parameters : dict, optional
+        Maps SymPy Symbol to numbers, for overriding the values of parameters in the Database.
+    callables : dict, optional
+        Pre-computed callables
+    output : str
+        Output property of the particular Model to sample
+    build_gradients : bool
+        Whether or not to build gradient functions. Defaults to True.
+
+    verbose : bool
+        Print the name of the phase when its callables are built
+
+    Returns
+    -------
+    callables : dict
+        Dictionary of keyword argument callables to pass to equilibrium.
+
+    Example
+    -------
+    >>> dbf = Database('AL-NI.tdb')
+    >>> comps = ['AL', 'NI', 'VA']
+    >>> phases = ['FCC_L12', 'BCC_B2', 'LIQUID', 'AL3NI5', 'AL3NI2', 'AL3NI']
+    >>> callables = build_callables(dbf, comps, phases)
+    >>> equilibrium(dbf, comps, phases, conditions, **callables)
+    """
+    parameters = parameters if parameters is not None else {}
+    if len(parameters) > 0:
+        param_symbols, param_values = zip(*[(key, val) for key, val in sorted(parameters.items(),
+                                                                              key=operator.itemgetter(0))])
+        param_values = np.asarray(param_values, dtype=np.float64)
+    else:
+        param_symbols = []
+        param_values = np.empty(0)
+    comps = sorted(unpack_components(dbf, comps))
+    pure_elements = get_pure_elements(dbf, comps)
+
+    callables = callables if callables is not None else {}
+    _callables = {
+        'massfuncs': {},
+        'massgradfuncs': {},
+        'callables': {},
+        'grad_callables': {},
+        'hess_callables': defaultdict(lambda: None),
```

Comment:
Can/should we remove the Hessian callables?

Reply:
Yes but I think its usage is annoyingly hardcoded into a few function signatures. Is it used in ESPEI?

Reply:
Everything in ESPEI just passes around the callables from `build_callables` now